js/ad_manager_vast.js

```diff
@@ -1875,6 +2127,625 @@ OO.Ads.manager(function(_, $) {
+      } else {
+        videoClickTracking = {
+          nonLinearClickThrough: _cleanString(this.$_node.find('NonLinearClickThrough').text())
+        }
+      }
+
+      var sequence = this.$_node.attr('sequence');
+      var adPodLength = this.$_node.parent().find('[sequence] Linear').length;
+      if (!supportsPoddedAds(version) || !_.isNumber(parseInt(sequence))) {
+        sequence = null;
+      }
+
+      var companionAds = [];
+      var companions = $companionsNode.find('Companion');
+      if (companions.length !== 0) {
+        companions.each(function(i, v){
+          companionAds.push(parseCompanionAd($(v)));
+        });
+      }
+      // this is for linear/nonlinear
+      var ad = {
+        mediaFiles: mediaFile,
+        tracking: tracking,
+        duration: isLinear ? this.$_node.find('Duration').text() : 0,
+        skipOffset: $node.attr('skipoffset') || null
+      };
+      _.extend(ad, videoClickTracking);
+
+      var data = {
+        adType: 'vpaid',
+        companion: companionAds,
+        error: errorTracking,
+        impression: impressions,
+        linear: ad,
+        nonLinear: ad,
+        title: _cleanString(this.$_node.find('AdTitle').text()),
+        tracking: tracking,
+        type: isLinear ? this.amc.ADTYPE.LINEAR_VIDEO : this.amc.ADTYPE.NONLINEAR_OVERLAY,
+        version: version,
+        videoClickTracking: videoClickTracking
+      };
+
+      var result = {
+        adPodIndex: parseInt(sequence) || 1,
+        sequence: sequence || null,
+        adPodLength: adPodLength ? adPodLength : 1,
+        data: data,
+        fallbackAd: null,
+        positionSeconds: adLoaded.time / 1000,
+        adParams: adParams,
+        streams: {'mp4': ''},
+        type: AD_TYPE.INLINE,
+        mediaFile: mediaFile,
+        version: version,
+        durationInMilliseconds: OO.timeStringToSeconds(ad.duration) * 1000
+      };
+
+      return result;
+    }, this);
+
+    /**
+     * Starts the click-to-linear ad
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_beginAd
+     */
+    var _beginAd = _.bind(function() {
+      var ad = currentAd.vpaidAd,
+          adLinear = ad.getAdLinear();
+
+      initSkipAdOffset(currentAd);
+      //Since IMA handles its own UI, we want the video player to hide its UI elements
+      this.amc.hidePlayerUi();
+      this.amc.notifyPodStarted(currentAd.id, currentAd.ad.adPodLength);
+
+      this.amc.notifyLinearAdStarted(currentAd.id, {
+        name: currentAd.data.title,
+        duration : ad.getAdDuration(),
+        clickUrl: _hasClickUrl(currentAd),
+        indexInPod: currentAd.ad.sequence,
+        skippable : ad.getAdSkippableState()
+      });
+    }, this);
+
+    /**
+     * Once Ad Playback stopped
+     * This is only required for VPAID ads
+     * @private
+     * @method Vast#_stopAd
+     */
+    var _stopAd = _.bind(function() {
+      if (currentAd && currentAd.vpaidAd) {
+        currentAd.vpaidAd.stopAd();
```

Comment:
For statements like this you could create a "safeFunctionCall" method. AMC has code like that and so does the AF,if you need an example.